[PROF-7359] Document workaround for Ruby profiler signal usage (#17370)
DataDog/documentation: content/en/profiler/profiler_troubleshooting.md

end
```

> ## Unexpected run-time failures and errors from Ruby gems that use native extensions in `dd-trace-rb` 1.11.0+

Reviewer: Should run-time be spelled runtime?

Reviewer: That makes sense. Thanks for the link.

Starting from `dd-trace-rb` 1.11.0, the "CPU Profiling 2.0" profiler gathers data by sending `SIGPROF` unix signals to Ruby applications, enabling finer-grained data gathering.

Sending `SIGPROF` is a common profiling approach, and may cause system calls from native extensions/libraries to be interrupted with a system [`EINTR` error code][8].

> Rarely, native extensions or libraries called by them may be have missing or incorrect error handling for the `EINTR` error code.

Reviewer: ```suggestion
Rarely, native extensions or libraries called by them may have missing or incorrect error handling for the `EINTR` error code.
```

> One known instance of this issue is when using the `mysql2` gem together with versions of libmysqlclient [older than 8.0.0][9]. The affected libmysqlclient version is known to be present on Ubuntu 18.04, but not 20.04 and later releases.
> **Note**: For this case, the profiler auto-detects when the `mysql2` gem is in use and auto-applies the solution described below.

Reviewer: ```suggestion
One known instance of this issue is when using the `mysql2` gem together with versions of libmysqlclient [older than 8.0.0][9]. The affected libmysqlclient version is known to be present on Ubuntu 18.04, but not 20.04 and later releases. For this case, the profiler auto-detects when the `mysql2` gem is in use and auto-applies the solution described below.
```

end
```

## Unexpected run-time failures and errors from Ruby gems that use native extensions in `dd-trace-rb` 1.11.0+

Starting from `dd-trace-rb` 1.11.0, the "CPU Profiling 2.0" profiler gathers data by sending `SIGPROF` unix signals to Ruby applications, enabling finer-grained data gathering.

> Sending `SIGPROF` is a common profiling approach, and may cause system calls from native extensions/libraries to be interrupted [with a system `EINTR` error code](https://man7.org/linux/man-pages/man7/signal.7.html#:~:text=Interruption%20of%20system%20calls%20and%20library%20functions%20by%20signal%20handlers).

Reviewer: ```suggestion
Sending `SIGPROF` is a common profiling approach, and may cause system calls from native extensions/libraries to be interrupted with a system [`EINTR` error code][4].
```

Rarely, native extensions or libraries called by them may be have missing or incorrect error handling for the `EINTR` error code.

> One known instance of this issue is when using the `mysql2` gem together with versions of libmysqlclient [older than 8.0.0](https://bugs.mysql.com/bug.php?id=83109). The affected libmysqlclient version is known to be present on Ubuntu 18.04, but not 20.04 and later releases.

Reviewer: ```suggestion
One known instance of this issue is when using the `mysql2` gem together with versions of libmysqlclient [older than 8.0.0][5]. The affected libmysqlclient version is known to be present on Ubuntu 18.04, but not 20.04 and later releases.
```

> **Note**: For this case, the profiler auto-detects when the `mysql2` gem is in use and auto-applies the solution described below.

Author: I'm a bit less sure about this suggestion. My intention is that the whole paragraph is supposed to sound like "We actually found an example of this issue... and we auto-detect this one, so don't worry". (And later on we ask customers to report these issues to us, so that we can extend this auto-detection)

I initially added this as a "Note that for this case...". The linter told me to add a `**Note**` instead, which already is kinda weird because it makes that part pop up in a weird way from the paragraph, but I went with it.

But completely separating the note from the paragraph I think breaks the connection I originally intended, and seems to make it an independent note. But then the note starts with "For this case", which is supposed to refer to the previous sentence.

So I think if we really go with the separation, we need to rewrite the "for this case as well". Any suggestions?

> **Note**: For this case, the profiler auto-detects when the `mysql2` gem is in use and auto-applies the solution described below.

Reviewer: Got it. That makes sense. I'll update my suggestion so that it remains one paragraph.

> If you run into run-time failures and/or errors from Ruby gems that use native extensions, you can revert back to the legacy profiler.
> The legacy profiler does not use `SIGPROF` signals.
> You can do this by setting the `DD_PROFILING_FORCE_ENABLE_LEGACY` environment variable to `true`, or in code:

Reviewer: ```suggestion
If you encounter run-time failures or errors from Ruby gems that use native extensions, you can revert back to the legacy profiler which does not use `SIGPROF` signals. To revert to the legacy profiler, set the `DD_PROFILING_FORCE_ENABLE_LEGACY` environment variable to `true`, or in code:
```

```ruby
Datadog.configure do |c|
  c.profiling.advanced.force_enable_legacy_profiler = true
end
```

Let our team know if you find (or suspect) any such incompatibilities [by opening a support ticket][2].
Doing this enables Datadog to add them to the auto-detection list, and to work with the gem/library authors to fix the issue.

[1]: /tracing/troubleshooting/#tracer-debug-logs
[2]: /help/

> [3]: https://github.com/DataDog/dd-trace-rb/releases/tag/v0.54.0

Reviewer: ```suggestion
[3]: https://github.com/DataDog/dd-trace-rb/releases/tag/v0.54.0
[4]: https://man7.org/linux/man-pages/man7/signal.7.html#:~:text=Interruption%20of%20system%20calls%20and%20library%20functions%20by%20signal%20handlers
[5]: https://bugs.mysql.com/bug.php?id=83109
```
